File: data structures/cpp/linked_list/linked_list.cpp
#include <iostream>
#include <memory>
#include <ostream>
#include <stdexcept>
#include <utility>

class Node {
 private:
  int data;
  std::unique_ptr<Node> next;

  // Give LinkedList access to private members
  friend class LinkedList;

 public:
  Node(int value) : data(value), next(nullptr) {}
};

class LinkedList {
 private:
  std::unique_ptr<Node> head;
  int size;

 public:
  LinkedList() : head(nullptr), size(0) {}

  void pushBack(int value) {
    // unique ptr
    auto newNode = std::make_unique<Node>(value);

    if (!head) {
      head = std::move(newNode);
    } else {
      Node* current = head.get();

      while (current->next != nullptr) {
        current = current->next.get();
      }

      current->next = std::move(newNode);
    }
    ++size;
  }

  void pushFront(int value) {
    auto newNode = std::make_unique<Node>(value);

    if (!head) {
      head = std::move(newNode);
    } else {
      newNode->next = std::move(this->head);
      this->head = std::move(newNode);
    }
    ++size;
  }

  void printList() const {
    if (!head) {
      return;
    } else {
      Node* current = head.get();

      while (current != nullptr) {
        std::cout << current->data;
        if (current->next != nullptr) {
          std::cout << ", ";
        }
        current = current->next.get();
      }
      std::cout << std::endl;
    }
  }

  int popBack() {
    if (!head) {
      throw std::runtime_error("Cannot pop from an empty list");
    } else {
      Node* current = head.get();

      auto prev = current;

      if (current->next == nullptr) {
        auto back = head->data;
        head = nullptr;
        --size;
        return back;
      }

      while (current->next != nullptr) {
        prev = current;
        current = current->next.get();
      }
      auto back = current->data;
      prev->next.reset();
      --size;
      return back;
    }
  }

  int popFront() {
    if (!head) {
      throw std::runtime_error("Cannot pop from an empty list");
    } else {
      if (head->next == nullptr) {
        auto front = head->data;
        head = nullptr;
        --size;
        return front;
      } else {
        auto front = head->data;
        head = std::move(head->next);
        --size;
        return front;
      }
    }
  }

  int front() const {
    if (!head) {
      throw std::runtime_error("Empty list!");
    }
    return head->data;
  }

  int back() const {
    if (!head) {
      throw std::runtime_error("Empty list!");
    }
    const Node* current = head.get();

    while (current->next != nullptr) {
      current = current->next.get();
    }
    return current->data;
  }

  bool empty() const { return !head; }
};

int main(void) {
  LinkedList list;

  list.pushFront(33);
  list.pushFront(50);
  list.pushFront(21);
  std::cout << "BACK: " << list.back() << std::endl;

  return 0;
}

File: data structures/python/linked_list/linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self, head, size):
        self.head = None
        self.size = 0

    # Adds the element `value` to the end of the linked list.
    def push_back(self, value):
        new_node = Node(value)
        if not new_node:
            return
        new_node.data = value
        new_node.next = None

        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            
            current.next = new_node
        
        self.size += 1

    # Inserts the element `value` at the beginning of the linked list
    def push_front(self, value):
        new_node = Node(value)
        if not new_node:
            return
        new_node.data = value
        new_node.next = self.head
        self.head = new_node
        self.size += 1

    def print_list(self):
        current = self.head
        while current is not None:
            print(current.data)




def main():
    linked_list = LinkedList()